Sources/Sentry/SentryViewHierarchy.m

```diff
@@ -1,48 +1,140 @@
 
+static int
+writeJSONDataToFile(const char *const data, const int length, void *const userData)
+{
+    const int fd = *((int *)userData);
+    const bool success = sentrycrashfu_writeBytesToFD(fd, data, length);
+    return success ? SentryCrashJSON_OK : SentryCrashJSON_ERROR_CANNOT_ADD_DATA;
+}
+
+static int
+writeJSONDataToMemory(const char *const data, const int length, void *const userData)
+{
+    NSMutableData *memory = ((__bridge NSMutableData *)userData);
+    [memory appendBytes:data length:length];
+    return SentryCrashJSON_OK;
+}
+
 @implementation SentryViewHierarchy
 
-- (NSArray<NSString *> *)fetchViewHierarchy
+- (BOOL)saveViewHierarchy:(NSString *)filePath
 {
-    return [self fetchViewHierarchyPreventMoveToMainThread:NO];
+    NSArray<UIWindow *> *windows = [SentryDependencyContainer.sharedInstance.application windows];
+
+    const char *path = [filePath UTF8String];
+    int fd = open(path, O_RDWR | O_CREAT | O_TRUNC, 0644);
+    if (fd < 0) {
+        SENTRY_LOG_DEBUG(@"Could not open file %s for writing: %s", path, strerror(errno));
+        return false;
+    }
+
+    BOOL result = [self processViewHierarchy:windows addFunction:writeJSONDataToFile userData:&fd];
+
+    close(fd);
+    return result;
 }
 
-- (NSArray<NSString *> *)fetchViewHierarchyPreventMoveToMainThread:(BOOL)preventMoveToMainThread
+- (NSData *)fetchViewHierarchy
 {
     NSArray<UIWindow *> *windows = [SentryDependencyContainer.sharedInstance.application windows];
 
-    NSMutableArray *result = [NSMutableArray arrayWithCapacity:[windows count]];
+    __block NSMutableData *result = [[NSMutableData alloc] init];
 
-    [windows enumerateObjectsUsingBlock:^(UIWindow *window, NSUInteger idx, BOOL *stop) {
-        // In the case of a crash we can't dispatch work to be executed anymore,
-        // so we'll run this on the wrong thread.
-        if ([NSThread isMainThread] || preventMoveToMainThread) {
-            [result addObject:[window sentry_recursiveViewHierarchyDescription]];
-        } else {
-            dispatch_sync(dispatch_get_main_queue(),
-                ^{ [result addObject:[window sentry_recursiveViewHierarchyDescription]]; });
+    void (^save)(void) = ^{
+        if (![self processViewHierarchy:windows
+                            addFunction:writeJSONDataToMemory
+                               userData:(__bridge void *)(result)]) {
+            result = nil;
         }
-    }];
+    };
+
+    // In the case of a crash we can't dispatch work to be executed anymore,
+    // so we'll run this on the wrong thread.
+    if ([NSThread isMainThread]) {
+        save();
+    } else {
+        dispatch_sync(dispatch_get_main_queue(), save);
+    }
 
     return result;
 }
 
-- (void)saveViewHierarchy:(NSString *)path
+#    define tryJson(code)                                                                          \
+        if ((result = (code)) != SentryCrashJSON_OK)                                               \
+        goto done
```

Comment:
`goto done`? This seems.. very old fashioned and strange to my eyes. Is this the best way to solve this?

Reply:
We already have this SentryCrashJsonEncoder, this is how it works, and it works pretty well. 
Doesn't make sense to have another JSON serialiser

Reply:
I strongly disagree, but ok.